openedx/openedx-platform
# feat: Course roles service

Description
This pull request adds the new CourseRoles model that will be used for creating roles with associated permissions that can be assigned to users on a course, as an org-wide course role (the user has access as if they have the role on all courses in the org), or instance-wide course role (the user has access as if they have the role on all courses within the instance).

This initial PR does not create any roles, but does setup the ability to do so in the future.

Supporting information
This PR is the initial work for the [Tech Spec related to Course Roles](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3824648277/RBAC+Tech+Spec).

Testing instructions
In order to run the code fully, a waffle flag (course_roles.use_permission_checks) should be created. If the waffle flag is not present or is false, the permissions checks will return false. Only if the waffle flag is true will the permissions checks check the db for access.

There is no intended change in functionality for a user. Testing should reflect no noticeable user experience changes as there are currently no course_roles_roles created and no users who are assigned the new course_roles_roles.

Other information

[openedx/core/djangoapps/course_roles/docs/decisions/0001-course_level_roles.rst](https://github.com/openedx/edx-platform/pull/33609/files/eab726e8a1267b4c0e31cbab0df6a7df9196c9bd#diff-09d45b3a47f61c47edabbca9774601e4de1887ef0b621084af693cd06ce62003) provides context on these changes and may be a good starting point for reviewing the PR

This was originally included as a part of a [larger draft PR](https://github.com/openedx/edx-platform/pull/33609) that is being split into multiple PRs.
